host/src/execution.rs

```diff
@@ -37,42 +37,40 @@ pub async fn execute<D: Prover>(
 
     // 2. Test run the block
     memory::reset_stats();
-    let build_result = TaikoStrategy::build_from(&input);
-    let output = match &build_result {
+    match TaikoStrategy::build_from(&input) {
         Ok((header, _mpt_node)) => {
             info!("Verifying final state using provider data ...");
             info!("Final block hash derived successfully. {}", header.hash());
             info!("Final block header derived successfully. {:?}", header);
-            let pi = D::instance_hash(assemble_protocol_instance(&input, header)?);
+            let pi = D::instance_hash(assemble_protocol_instance(&input, &header)?);
             // Make sure the blockhash from the node matches the one from the builder
             assert_eq!(header.hash().0, input.block_hash, "block hash unexpected");
-            GuestOutput::Success((
+            let output = GuestOutput::Success((
                 WrappedHeader {
                     header: header.clone(),
                 },
                 pi,
-            ))
+            ));
+            memory::print_stats("Guest program peak memory used: ");
+
+            // Prove
```

Comment:
Any specific reason you moved the prove in here?

Reply:
I think no need to run prover if build fails?
Previous logic is if build from fails, it sets output to failure and still run prove with a invalid output, it looks strange.

Reply:
Fair point! I guess normally it should never be possible to make it impossible to prove a block, so I'm sometimes wondering when that would actually happen. But it certainly can if there is a bug somewhere in the input generation or something like that.

Reply:
Maybe it is related to invalid block case?
if so, I can revert it.

Reply:
All invalid bock cases still have to be successful because they do indeed have to proven. So the reason why invalid is return should be either a bug or some unsupported feature, which indeed should then skip the proving so I do think your change is correct!